devops/clean

```diff
@@ -0,0 +1,66 @@
+#!/bin/bash
+# Cleans all build artifacts and custom config from the repo.
+# Intended for use on developer workstations. Should NOT be run
+# on production Admin Workstations.
+set -e
+set -u
+
+
+function remove_unwanted_files() {
+
+    # Remove any Onion URL info from previous instances.
+    rm -vf install_files/ansible-base/app-source-ths \
+        install_files/ansible-base/app-ssh-aths \
+        install_files/ansible-base/app-journalist-aths \
+        install_files/ansible-base/mon-ssh-aths \
+        build/*.deb
+
+    # Remove extraneous copies of the git repos, pulled in
+    # via the Molecule upgrade testing scenario.
+    rm -rf molecule/upgrade/.molecule/sd-orig \
+        molecule/vagrant_packager/.molecule/sd-orig
+
+    # Python bytecode, left over from tests or local app runs.
+    find "$PWD" -type f -iname '*.pyc' -delete
+
+    # Ansible playbook retry files, which are never used.
+    find "$PWD" -type f -iname '*.retry' -delete
+
+    # Static assests generated by local dev, as well as
+    # virtualenv for securedrop-admin script.
+    rm -rvf securedrop/static/.webassets-cache .venv
+
+    # Any Vagrant VMs should be destroyed. We check for the .vagrant/
+    # directory, because it'll be created by vagrant otherwise, breaking
+    # idempotence for the cleanup action.
+    if hash vagrant &> /dev/null && [[ -d .vagrant/ ]]
+    then
+        printf "Removing vagrant VMs\\n"
+        vagrant destroy -f > /dev/null
```

Comment:
So my initial recommendation was a little off. `vagrant destroy -f` returns non-zero if there are no live VMs (which is counter to what `-f` mean for most remove commands). This needs to be guarded with `|| echo 'No VMs removed.'` perhaps.

Reply:
@heartsucker The conditional block avoids the non-zero exit, since it checks for existence of the `.vagrant/` dir. In general I'm a fan of verbose scripts, but the rest of the script prints deletions if-and-only-if they occurred, so I'm inclined to omit further changes here in the interest of consistency.

Reply:
Hmm even with `set -x`, I get some failures.

```
$ vagrant up /staging/ --no-provision && make clean
Bringing machine 'mon-staging' up with 'virtualbox' provider...
Bringing machine 'app-staging' up with 'virtualbox' provider...
==> mon-staging: Importing base box 'bento/ubuntu-14.04'...
==> mon-staging: Matching MAC address for NAT networking...
==> mon-staging: Checking if box 'bento/ubuntu-14.04' is up to date...
==> mon-staging: A newer version of the box 'bento/ubuntu-14.04' is available! You currently
==> mon-staging: have version '201803.24.0'. The latest is version '201806.08.0'. Run
==> mon-staging: `vagrant box update` to update.
==> mon-staging: Setting the name of the VM: securedrop_mon-staging_1530607282211_90442
==> mon-staging: Clearing any previously set network interfaces...
==> mon-staging: Preparing network interfaces based on configuration...
    mon-staging: Adapter 1: nat
    mon-staging: Adapter 2: hostonly
==> mon-staging: Forwarding ports...
    mon-staging: 22 (guest) => 2222 (host) (adapter 1)
==> mon-staging: Booting VM...
==> mon-staging: Waiting for machine to boot. This may take a few minutes...
    mon-staging: SSH address: 127.0.0.1:2222
    mon-staging: SSH username: vagrant
    mon-staging: SSH auth method: private key
==> mon-staging: Machine booted and ready!
==> mon-staging: Checking for guest additions in VM...
==> mon-staging: Setting hostname...
==> mon-staging: Configuring and enabling network interfaces...
==> mon-staging: Machine not provisioned because `--no-provision` is specified.
==> app-staging: Importing base box 'bento/ubuntu-14.04'...
==> app-staging: Matching MAC address for NAT networking...
==> app-staging: Checking if box 'bento/ubuntu-14.04' is up to date...
==> app-staging: A newer version of the box 'bento/ubuntu-14.04' is available! You currently
==> app-staging: have version '201803.24.0'. The latest is version '201806.08.0'. Run
==> app-staging: `vagrant box update` to update.
==> app-staging: Setting the name of the VM: securedrop_app-staging_1530607313654_20237
==> app-staging: Fixed port collision for 22 => 2222. Now on port 2200.
==> app-staging: Clearing any previously set network interfaces...
==> app-staging: Preparing network interfaces based on configuration...
    app-staging: Adapter 1: nat
    app-staging: Adapter 2: hostonly
==> app-staging: Forwarding ports...
    app-staging: 22 (guest) => 2200 (host) (adapter 1)
==> app-staging: Running 'pre-boot' VM customizations...
==> app-staging: Booting VM...
==> app-staging: Waiting for machine to boot. This may take a few minutes...
    app-staging: SSH address: 127.0.0.1:2200
    app-staging: SSH username: vagrant
    app-staging: SSH auth method: private key
==> app-staging: Machine booted and ready!
==> app-staging: Checking for guest additions in VM...
==> app-staging: Setting hostname...
==> app-staging: Configuring and enabling network interfaces...
==> app-staging: Machine not provisioned because `--no-provision` is specified.
+ confirmation_prompt
+ cat
WARNING: Proceeding will destroy all local customizations, including
application configuration, user credentials, Onion URLs, static assets,
VMs, and all git-ignored files. It is NOT possible to undo these changes!
+ read -r -p 'To proceed, type '\''yes'\'': ' user_confirmation
To proceed, type 'yes': yes
+ case "${user_confirmation}" in
+ remove_unwanted_files
+ rm -vf install_files/ansible-base/app-source-ths install_files/ansible-base/app-ssh-aths install_files/ansible-base/app-journalist-aths install_files/ansible-base/mon-ssh-aths 'build/*.deb'
+ rm -rf molecule/upgrade/.molecule/sd-orig molecule/vagrant_packager/.molecule/sd-orig
+ find /home/heartsucker/code/freedomofpress/securedrop -type f -iname '*.pyc' -delete
+ find /home/heartsucker/code/freedomofpress/securedrop -type f -iname '*.retry' -delete
+ rm -rvf securedrop/static/.webassets-cache .venv
+ hash vagrant
+ [[ -d .vagrant/ ]]
+ printf 'Removing vagrant VMs\n'
Removing vagrant VMs
+ vagrant destroy -f
==> app-prod: VM not created. Moving on...
==> mon-prod: VM not created. Moving on...
==> app-staging: Forcing shutdown of VM...
==> app-staging: Destroying VM and associated drives...
==> mon-staging: Forcing shutdown of VM...
==> mon-staging: Destroying VM and associated drives...
Makefile:152: recipe for target 'clean' failed
make: *** [clean] Error 2
```

Reply:
@heartsucker I'm unable to reproduce your results:

```
 $ vagrant up /staging/ --no-provision && make clean
Bringing machine 'mon-staging' up with 'libvirt' provider...
Bringing machine 'app-staging' up with 'libvirt' provider...
==> app-staging: Creating image (snapshot of base box volume).
==> mon-staging: Creating image (snapshot of base box volume).
==> app-staging: Creating domain with the following settings...
==> mon-staging: Creating domain with the following settings...
[...snip...]
WARNING: Proceeding will destroy all local customizations, including
application configuration, user credentials, Onion URLs, static assets,
VMs, and all git-ignored files. It is NOT possible to undo these changes!
To proceed, type 'yes': yes
Removing vagrant VMs
Removing .vagrant/
 $ vagrant --version
Vagrant 1.9.1
```

Out of curiosity, which version of vagrant are you using? I don't mind adding the `||` you suggest! Just trying to figure out the cause, since if we have divergence, we should probably pin a version of vagrant to support (cf. #3350). 

N.B. the [vagrant changelog](https://github.com/hashicorp/vagrant/blob/04f7215b5e3f594368696262920deaf631315e37/CHANGELOG.md#202-january-29-2018) notes that in 2.0.1, the exit code for destroy changed:

> commands/destroy: Exit 0 if vagrant destroy finds no running vms [GH-9251]

but that description doesn't match what you're seeing.

Reply:
So on vagrant 2.0.1, I saw the same as @heartsucker: 

```
$ make clean
WARNING: Proceeding will destroy all local customizations, including
application configuration, user credentials, Onion URLs, static assets,
VMs, and all git-ignored files. It is NOT possible to undo these changes!
To proceed, type 'yes': yes
Removing vagrant VMs
make: *** [clean] Error 2
```

Which confused me. So I upgraded to latest vagrant, which should be fine since the changelog indicates the exit codes were fixed after 2.0.2 (which apparently was the version that modified the exit code behavior).

Unfortunately, it turns out that even on latest, if I have a vagrant VM up and i `make clean`, it will still exit 2, due to the logic [here](https://github.com/hashicorp/vagrant/blob/master/plugins/commands/destroy/command.rb#L70), where vagrant destroy still returns 2 (see the bugfix they applied in 2.0.2 here: https://github.com/hashicorp/vagrant/pull/9251/files which doesn't fix this). I filed https://github.com/hashicorp/vagrant/issues/10024 to figure out if this is a bug or not. 

In the meantime, what about removing `set -e`?

Reply:
Great debugging, @redshiftzero ! How frustrating. 

> In the meantime, what about removing set -e?

I'd vastly prefer @heartsucker's suggestion of `|| true` simply on the `vagrant destroy` command. Acceptable?

Reply:
Honestly, I don't really have any strong feelings about this as long as `make clean` does not produce an error when:

1. No VMs are deleted (@heartsucker's first case)
2. One VM is deleted (my case)

The guard proposed would address case one, but produce inaccurate output if the suggested `|| echo 'No VMs removed.'` text is used, so as long as we change that text to be more general I'm cool with it

Reply:
Try again, @redshiftzero ? I revised the logic to ignore errors, and made the messaging more general. For me, the following are true:

* `make clean` _with_ VMs displays "Removing VMs... done" with no error
* `make clean` _without_ VMs displays nothing, with no error.

Again, I'm on vagrant 1.9.1. If you see different results, happy to investigate further.

Reply:
(Verified this behavior in both these cases :+1:)